### PR TITLE
Refactor implementations of query phase searcher, add empty QueryCollectorContext

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description
Current implementations of query phase searcher (default and concurrent) alway adds `TopDocsCollector` collector to the collector context. In some scenario it's not the preferred behavior, for instance if in a plugin we're adding custom collector that collects doc scores in a different fashion. With current implementation there will be two collectors in the context, this will create a performance overhead. 

With this change we're not changing core functionality or implementation of query phase searcher. Instead we're adding overloaded `searchWith` method that child query phase search can override and pass empty `QueryPhaseSearcher`.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/13170

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by martin-gaievski** [comment]
this change need to be backported to 2.x, need help from maintainers, I don't have permissions to add a label

**Comment by martin-gaievski** [comment]
@reta can you please review this change, this is related to https://github.com/opensearch-project/OpenSearch/pull/13481? I've tested the PR with corresponding changes at the plugin level, I can created PR for neural-search repo once core change is merged.

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:63]: This will not work,  ideally the `shouldRescore` should be invoked **after** the search finishes since it may use whatever state it may have accumulated

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:134]: We may need to add the interface here:

```
interface RescoringQueryCollectorContext {
    boolean shouldRescore();
}
```

```
if (queryCollectorContext instanceof RescoringQueryCollectorContext) {
    return ((RescoringQueryCollectorContext) queryCollectorContext).shouldRescore();
}
```

File: server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java:252]: Under which conditions `in` is `null`?

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:99]: Should we use `EMPTY_CONTEXT` instead of dealing with `null`s?

```
    public static final QueryCollectorContext EMPTY_CONTEXT = new QueryCollectorContext("empty") {
        @Override
        Collector create(Collector in) throws IOException {
            return EMPTY_COLLECTOR;
        }

        @Override
        CollectorManager<?, ReduceableSearchResult> createManager(CollectorManager<?, ReduceableSearchResult> in) throws IOException {
            return new CollectorManager<Collector, ReduceableSearchResult>() {
                @Override
                public Collector newCollector() throws IOException {
                    return EMPTY_COLLECTOR;
                }

                @Override
                public ReduceableSearchResult reduce(Collection<Collector> collectors) throws IOException {
                    return result -> {};
                }
            };
        }
    };
```

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:134]: makes sense, will add the interface. I assume that the default value (if the passed context is not instance of our new interface) is `false`

File: server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java:252]: As per logic in [QueryCollectorManagerContext](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/query/QueryCollectorManagerContext.java#L71) this is `null` when we iterate over collection of collectorContexts and this is first collectContext in that collection.
In my test that happened if I'm skip adding the TopDocsCollector

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:99]: let me try this approach, it would be nice if that works, I don't like dealing with `null` either

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:63]: got it, adding `RescoringQueryCollectorContext` interface should solve this

Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:63]: Added `RescoringQueryCollectorContext` interface

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:134]: Added `RescoringQueryCollectorContext` interface

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:99]: Added EMPTY_CONTEXT static instance

File: server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java]: I've removed this check

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:97]: I don't think you need `shouldRescore` anymore

Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java]: Yup, removed `shouldRescore` argument but forgot to add to commit. Now it should be fixed

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:87]: I think we should not expect `null` anymore but `QueryCollectorContext.EMPTY_CONTEXT`?

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:58]: Btw, I think this method should be part of the `DefaultQueryPhaseSearcher` right? (not only concurrent one)

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java]: I mean replaced `if` for nonNull condition by  nonNull `assert`

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (reta) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:58]: Oh sorry, `@Override` is missed

Comment (reta) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:58]:
```
@Override
protected boolean searchWithCollector(
```

**Review by reta — APPROVED** [review]
LGTM, thanks @martin-gaievski , @jed326 do you mind to cross check this one please? thank you!

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (jed326) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:58]: I think there is an `@Override` missing here. Also can we just call `searchWithCollectorManager` directly on L55 above?

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (jed326) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:56]: It looks like this is identical as what is in the method being overridden here?

File: server/src/main/java/org/opensearch/search/query/QueryPhase.java
Comment (jed326) [inline server/src/main/java/org/opensearch/search/query/QueryPhase.java:450]: Why do we need this new method instead of just calling `QueryPhase.searchWithCollector` on L447 above?

Also having 3 `searchWithCollector` implementations that call and override each other is quite confusing to read although I don't have any suggestions for a better name right now

File: server/src/main/java/org/opensearch/search/query/QueryPhase.java
Comment (jed326) [inline server/src/main/java/org/opensearch/search/query/QueryPhase.java:448]: Just for my understanding -- if I want to override the `TopDocsCollectorContext` at the top of the collector tree then with this change the way to do that is provide by own `QueryPhaseSearcher` that overrides this `searchWithCollector` method, right?

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:58]: Will add @Override in commit. We have to call new overloaded method, this allows to child implementation to override and pass empty context if needed. That was the main point for this change

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:56]: right, before this change we always create TopDocsCollector inside the `searchWithCollectorManager`. In some scenarios with don't need that collector (we are adding custom one instead), but because it's added it consumes resources. For detailed profiling info please check main GH issue.

File: server/src/main/java/org/opensearch/search/query/QueryPhase.java
Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/QueryPhase.java:448]: Yes, this is what we're going to do for hybrid query with [custom collector](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcher.java) in neural-search plugin. After this PR is merged I'll create one more PR where this change will be used, will ping you in that new PR.

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java
Comment (jed326) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:56]: I guess what I meant is why do we need this override at all then since we can use the `DefaultQueryPhaseSearcher` implementation in both cases? Can we just delete this method from `ConcurrentQueryPhaseSearcher`?

Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:56]: discussed offline, I missed the point that Concurrent QPS extends the DefaultQPS, so the existing override is same between parent and child classes, thus it's not needed. Remove the override method from ConcurrentQPS

**Review by jed326 — APPROVED** [review]

**Comment by martin-gaievski** [comment]
@reta can you please help with merging this PR, Jay has approved the PR

**Comment by reta** [comment]
> @reta can you please help with merging this PR, Jay has approved the PR

@martin-gaievski could you rebase p

> @reta can you please help with merging this PR, Jay has approved the PR

Fixing tests with https://github.com/opensearch-project/OpenSearch/pull/13510

**Comment by reta** [comment]
@martin-gaievski could you please rebase? thank you!

**Review by reta — APPROVED** [review]

**Pull request merged**
